content/posts/vault-migration/index.md

```diff
@@ -0,0 +1,558 @@
+
+To allow the restored Vault data to function properly, we need to configure the
+new instance to use the same unseal key as the old Vault. This step modifies the
+config map and restarts the pods to apply the change.
+
+- Patch the configuration of the new Vault cluster with the credentials of the
+  old Azure Key Vault
+
+  ```bash
+  kubectl get cm vault-config -o yaml > "${vault_config_filename}"
+
+  ed '+/seal "/' "${vault_config_filename}" <<EOF
+  .,/}/d
+  -
+  a
+      seal "azurekeyvault" {
+        tenant_id      = "${TENANT_ID}"
+        client_id      = "${NEW_VAULT_MSI_CLIENT_ID}"
+        vault_name     = "${OLD_KEY_VAULT_NAME}"
+        key_name       = "${OLD_VAULT_UNSEAL_KEY_SECRET_NAME}"
+      }
+  .
+  w
+  q
+  EOF
+
+  kubectl apply -f "${vault_config_filename}"
+  ```
+
+- Restart all pods to pick up the patched configuration
+  ```bash
+  kubectl delete pod --selector=app.kubernetes.io/name=vault
+  ```
+
+### 7. Prepare the unseal key migration on the new Vault instance
+
+To enable unseal key migration, we first need to configure the new Vault cluster
+to support dual seal configuration — keeping the old seal block active but
+disabled, and adding a new one that points to the new Azure Key Vault.
+
+- Update the Vault configuration:
+
+```bash
+kubectl get cm vault-config -o yaml > vault-config.yaml
+
+ed '+/seal "/' vault-config.yaml <<EOF
+a
+      disabled       = true
+.
+/}/
+a
+    seal "azurekeyvault" {
+      tenant_id      = "${TENANT_ID}"
+      client_id      = "${NEW_VAULT_MSI_CLIENT_ID}"
+      vault_name     = "${NEW_KEY_VAULT_NAME}"
+      key_name       = "${NEW_VAULT_UNSEAL_KEY_SECRET_NAME}"
+    }
+.
+w
+q
+EOF
+
+kubectl apply -f vault-config.yaml
+```
+
+### 8. Perform the unseal key migration on the follower pods
+
+Vault only allows unseal key migration on follower nodes, so we restart each
+follower pod and run the migration steps.
+
+- Iterate through each follower pod and migrate using all three unseal keys:
+
+  ```bash
+  for i in {0..2}; do
+    if [ "${i}" -eq "${leader_index}" ]; then
+      continue
+    fi
+
+    print_message "Restarting pod vault-${i} to start unseal key migration"
+    kubectl delete pod "vault-${i}"
+    kubectl wait \
+      --for=jsonpath='{.status.containerStatuses[0].state.running}' \
+      --selector=apps.kubernetes.io/pod-index="${i}" \
+      --timeout=90s \
+      pod
+
+    print_message "Migrating vault-${i}"
+    kubectl exec -it "vault-${i}" -- \
+      sh -c "vault operator unseal -migrate \"${OLD_VAULT_UNSEAL_KEY_1}\""
+    kubectl exec -it "vault-${i}" -- \
+      sh -c "vault operator unseal -migrate \"${OLD_VAULT_UNSEAL_KEY_2}\""
+    kubectl exec -it "vault-${i}" -- \
+      sh -c "vault operator unseal -migrate \"${OLD_VAULT_UNSEAL_KEY_3}\""
+  done
+  ```
+
+### 9. Step down leader pod
+
+Once all follower pods have been migrated, the leader must step down so it can
+restart and apply the new seal configuration.
```

Comment:
```suggestion
restart and apply the new seal configuration.\
Note that we use the migration token of the _old_ Vault instance here, since that is the token stored in the restored snapshot.
```